File: trade_analysis_web/src/strategy_core/local_fenxing_segment/momentum_exhaustion.ts
import { getBaseSegmentHighPoint, getBaseSegmentLowPoint } from './base_segment_builder'
import { getOffsetFromCurrentBar } from './base_fenxing_builder'
import type {
  BaseSegment,
  FenxingBar,
  MomentumExhaustionBuildState,
  MomentumExhaustionSignal,
  SegmentDirection,
} from './types'

type MacdCrossType = 'golden' | 'dead'

type SegmentRange = {
  high: number
  low: number
}

const isGoldenCross = (previousBar: FenxingBar, currentBar: FenxingBar) => {
  return previousBar.macdDiff <= previousBar.macdDea && currentBar.macdDiff > currentBar.macdDea
}

const isDeadCross = (previousBar: FenxingBar, currentBar: FenxingBar) => {
  return previousBar.macdDiff >= previousBar.macdDea && currentBar.macdDiff < currentBar.macdDea
}

const getSegmentClosedInterval = (segment: BaseSegment) => {
  return {
    endIndex: Math.max(segment.start.index, segment.end.index),
    startIndex: Math.min(segment.start.index, segment.end.index),
  }
}

const getRecentThreeSegmentsBeforeIndex = (
  baseSegments: BaseSegment[],
  barIndex: number,
) => {
  const availableSegments = baseSegments.filter((segment) => {
    const { endIndex } = getSegmentClosedInterval(segment)
    return endIndex <= barIndex
  })

  if (availableSegments.length < 3) {
    return null
  }

  return availableSegments.slice(-3)
}

const getSegmentRange = (segment: BaseSegment): SegmentRange => {
  return {
    high: getBaseSegmentHighPoint(segment).price,
    low: getBaseSegmentLowPoint(segment).price,
  }
}

const getSameDirectionMacdArea = (
  bars: FenxingBar[],
  segment: BaseSegment,
) => {
  const { startIndex, endIndex } = getSegmentClosedInterval(segment)
  let area = 0

  for (let index = startIndex; index <= endIndex; index += 1) {
    const bar = bars[index]
    if (!bar) {
      continue
    }

    if (segment.direction === 'up') {
      area += Math.max(bar.macdHistogram, 0)
      continue
    }

    area += Math.abs(Math.min(bar.macdHistogram, 0))
  }

  return area
}

const isValidThreeSegmentStructure = (
  segments: BaseSegment[],
  crossType: MacdCrossType,
) => {
  const [segmentA, segmentB, segmentC] = segments
  if (!segmentA || !segmentB || !segmentC) {
    return false
  }

  if (crossType === 'golden') {
    return segmentA.direction === 'down'
      && segmentB.direction === 'up'
      && segmentC.direction === 'down'
  }

  return segmentA.direction === 'up'
    && segmentB.direction === 'down'
    && segmentC.direction === 'up'
}

const isUpMomentumExhausted = (
  rangeA: SegmentRange,
  rangeC: SegmentRange,
  areaA: number,
  areaC: number,
) => {
  if (rangeA.low <= rangeC.low && rangeA.high < rangeC.high) {
    return areaC < areaA
  }

  if (rangeA.low <= rangeC.low && rangeA.high >= rangeC.high) {
    return true
  }

  if (rangeA.low > rangeC.low && rangeA.high < rangeC.high) {
    return false
  }

  return false
}

const isDownMomentumExhausted = (
  rangeA: SegmentRange,
  rangeC: SegmentRange,
  areaA: number,
  areaC: number,
) => {
  if (rangeA.high >= rangeC.high && rangeA.low > rangeC.low) {
    return areaC < areaA
  }

  if (rangeA.high >= rangeC.high && rangeA.low <= rangeC.low) {
    return true
  }

  if (rangeA.high < rangeC.high && rangeA.low > rangeC.low) {
    return false
  }

  return false
}

const buildSignalFromCross = (
  bars: FenxingBar[],
  crossBarIndex: number,
  direction: SegmentDirection,
  previousArea: number,
  currentArea: number,
): MomentumExhaustionSignal | null => {
  const crossBar = bars[crossBarIndex]
  if (!crossBar) {
    return null
  }

  return {
    currentStrength: currentArea,
    direction,
    point: {
      index: crossBar.index,
      price: direction === 'up' ? crossBar.high : crossBar.low,
      time: crossBar.time,
    },
    previousStrength: previousArea,
  }
}

const appendSignalForBar = (
  bars: FenxingBar[],
  baseSegments: BaseSegment[],
  barIndex: number,
) => {
  const previousBar = bars[barIndex - 1]
  const currentBar = bars[barIndex]
  if (!previousBar || !currentBar) {
    return null
  }

  let crossType: MacdCrossType | null = null
  if (isGoldenCross(previousBar, currentBar)) {
    crossType = 'golden'
  } else if (isDeadCross(previousBar, currentBar)) {
    crossType = 'dead'
  }

  if (!crossType) {
    return null
  }

  const recentSegments = getRecentThreeSegmentsBeforeIndex(baseSegments, currentBar.index)
  if (!recentSegments || !isValidThreeSegmentStructure(recentSegments, crossType)) {
    return null
  }

  const [segmentA, _, segmentC] = recentSegments
  if (!segmentA || !segmentC) {
    return null
  }

  const areaA = getSameDirectionMacdArea(bars, segmentA)
  const areaC = getSameDirectionMacdArea(bars, segmentC)
  const rangeA = getSegmentRange(segmentA)
  const rangeC = getSegmentRange(segmentC)

  if (crossType === 'dead') {
    if (!isUpMomentumExhausted(rangeA, rangeC, areaA, areaC)) {
      return null
    }

    return buildSignalFromCross(bars, currentBar.index, 'up', areaA, areaC)
  }

  if (!isDownMomentumExhausted(rangeA, rangeC, areaA, areaC)) {
    return null
  }

  return buildSignalFromCross(bars, currentBar.index, 'down', areaA, areaC)
}

export const createEmptyMomentumExhaustionBuildState = (): MomentumExhaustionBuildState => ({
  processedBarCount: 0,
  signals: [],
})

export const advanceMomentumExhaustionState = (
  buildState: MomentumExhaustionBuildState,
  bars: FenxingBar[],
  baseSegments: BaseSegment[],
) => {
  let latestSignal: MomentumExhaustionSignal | null = null
  const startBarIndex = Math.max(1, buildState.processedBarCount)

  for (let barIndex = startBarIndex; barIndex < bars.length; barIndex += 1) {
    const signal = appendSignalForBar(bars, baseSegments, barIndex)
    if (!signal) {
      continue
    }

    buildState.signals.push(signal)
    latestSignal = signal
  }

  buildState.processedBarCount = bars.length
  return latestSignal
}

export const rebuildMomentumExhaustionState = (
  bars: FenxingBar[],
  baseSegments: BaseSegment[],
) => {
  const buildState = createEmptyMomentumExhaustionBuildState()
  advanceMomentumExhaustionState(buildState, bars, baseSegments)
  return buildState
}

export const buildMomentumExhaustionSignals = (
  bars: FenxingBar[],
  baseSegments: BaseSegment[],
) => {
  return rebuildMomentumExhaustionState(bars, baseSegments).signals.map((signal) => ({
    ...signal,
    point: { ...signal.point },
  }))
}

export const getAllMomentumExhaustionSignals = (buildState: MomentumExhaustionBuildState) => {
  return buildState.signals.map((signal) => ({
    ...signal,
    point: { ...signal.point },
  }))
}

export const getMomentumExhaustionOutput = (
  bars: FenxingBar[],
  signals: MomentumExhaustionSignal[],
) => {
  const latestSignal = signals[signals.length - 1]

  if (!latestSignal) {
    return [Number.NaN, Number.NaN, Number.NaN, Number.NaN]
  }

  const offset = getOffsetFromCurrentBar(bars, latestSignal.point)

  if (latestSignal.direction === 'up') {
    return [latestSignal.point.price, offset, Number.NaN, Number.NaN]
  }

  return [Number.NaN, Number.NaN, latestSignal.point.price, offset]
}

File: trade_analysis_backend/app/services/analysis_core_v2/momentum_exhaustion.py
from dataclasses import dataclass, field

from .types import AnalysisBar, BaseSegment, MomentumExhaustionSignal, SegmentPoint


@dataclass
class MomentumExhaustionState:
    signals: list[MomentumExhaustionSignal] = field(default_factory=list)
    processed_bar_count: int = 0


def create_momentum_exhaustion_state() -> MomentumExhaustionState:
    return MomentumExhaustionState()


def _is_golden_cross(prev: AnalysisBar, curr: AnalysisBar) -> bool:
    if prev.macd_diff is None or prev.macd_dea is None:
        return False
    if curr.macd_diff is None or curr.macd_dea is None:
        return False
    return prev.macd_diff <= prev.macd_dea and curr.macd_diff > curr.macd_dea


def _is_dead_cross(prev: AnalysisBar, curr: AnalysisBar) -> bool:
    if prev.macd_diff is None or prev.macd_dea is None:
        return False
    if curr.macd_diff is None or curr.macd_dea is None:
        return False
    return prev.macd_diff >= prev.macd_dea and curr.macd_diff < curr.macd_dea


def _seg_closed_interval(seg: BaseSegment) -> tuple[int, int]:
    return min(seg.start.index, seg.end.index), max(seg.start.index, seg.end.index)


def _recent_three(
    base_segments: list[BaseSegment],
    bar_index: int,
) -> list[BaseSegment] | None:
    available = [seg for seg in base_segments if _seg_closed_interval(seg)[1] <= bar_index]
    if len(available) < 3:
        return None
    return available[-3:]


def _macd_area(bars: list[AnalysisBar], seg: BaseSegment) -> float:
    start_index, end_index = _seg_closed_interval(seg)
    area = 0.0
    for i in range(start_index, end_index + 1):
        histogram = bars[i].macd_histogram if i < len(bars) else None
        if histogram is None:
            continue
        if seg.direction == "up":
            area += max(histogram, 0.0)
        else:
            area += abs(min(histogram, 0.0))
    return area


def _seg_high(seg: BaseSegment) -> float:
    return max(seg.start.price, seg.end.price)


def _seg_low(seg: BaseSegment) -> float:
    return min(seg.start.price, seg.end.price)


def _is_up_exhausted(
    a: BaseSegment,
    c: BaseSegment,
    area_a: float,
    area_c: float,
) -> bool:
    a_low = _seg_low(a)
    a_high = _seg_high(a)
    c_low = _seg_low(c)
    c_high = _seg_high(c)

    if a_low <= c_low and a_high < c_high:
        return area_c < area_a
    if a_low <= c_low and a_high >= c_high:
        return True
    if a_low > c_low and a_high < c_high:
        return False
    return False


def _is_down_exhausted(
    a: BaseSegment,
    c: BaseSegment,
    area_a: float,
    area_c: float,
) -> bool:
    a_low = _seg_low(a)
    a_high = _seg_high(a)
    c_low = _seg_low(c)
    c_high = _seg_high(c)

    if a_high >= c_high and a_low > c_low:
        return area_c < area_a
    if a_high >= c_high and a_low <= c_low:
        return True
    if a_high < c_high and a_low > c_low:
        return False
    return False


def _check_current_bar(
    bars: list[AnalysisBar],
    base_segments: list[BaseSegment],
    index: int,
) -> MomentumExhaustionSignal | None:
    prev = bars[index - 1]
    curr = bars[index]

    cross_type: str | None = None
    if _is_golden_cross(prev, curr):
        cross_type = "golden"
    elif _is_dead_cross(prev, curr):
        cross_type = "dead"

    if cross_type is None:
        return None

    recent = _recent_three(base_segments, curr.index)
    if recent is None:
        return None

    a, b, c = recent

    if cross_type == "golden":
        if not (a.direction == "down" and b.direction == "up" and c.direction == "down"):
            return None
    else:
        if not (a.direction == "up" and b.direction == "down" and c.direction == "up"):
            return None

    area_a = _macd_area(bars, a)
    area_c = _macd_area(bars, c)

    if cross_type == "dead":
        if not _is_up_exhausted(a, c, area_a, area_c):
            return None
        return MomentumExhaustionSignal(
            direction="up",
            point=SegmentPoint(
                merged_index=curr.index,
                index=curr.index,
                price=curr.high,
                time=curr.time,
            ),
            previous_strength=area_a,
            current_strength=area_c,
        )

    if not _is_down_exhausted(a, c, area_a, area_c):
        return None
    return MomentumExhaustionSignal(
        direction="down",
        point=SegmentPoint(
            merged_index=curr.index,
            index=curr.index,
            price=curr.low,
            time=curr.time,
        ),
        previous_strength=area_a,
        current_strength=area_c,
    )


def advance_momentum_exhaustion(
    state: MomentumExhaustionState,
    bars: list[AnalysisBar],
    base_segments: list[BaseSegment],
    current_bar_index: int,
) -> None:
    if current_bar_index < 1 or current_bar_index < state.processed_bar_count:
        return

    signal = _check_current_bar(bars, base_segments, current_bar_index)
    if signal is not None:
        state.signals.append(signal)

    state.processed_bar_count = current_bar_index + 1
